batcher/aligned-batcher/src/retry.rs

```diff
@@ -0,0 +1,317 @@
+        .user_balances(*addr)
+        .call()
+        .await
+        .map_err(|e| {
+            warn!("Failed to get balance for address {:?}. Error: {e}", addr);
+            RetryError::Transient(e.to_string())
+        })
+}
+
+pub async fn get_user_nonce_from_ethereum_retryable(
+    payment_service: &BatcherPaymentService,
+    payment_service_fallback: &BatcherPaymentService,
+    addr: Address,
+) -> Result<U256, RetryError<String>> {
+    if let Ok(nonce) = payment_service.user_nonces(addr).call().await {
+        return Ok(nonce);
+    }
+    payment_service_fallback
+        .user_nonces(addr)
+        .call()
+        .await
+        .map_err(|e| {
+            warn!("Error getting user nonce: {e}");
+            RetryError::Transient(e.to_string())
+        })
+}
+
+pub async fn user_balance_is_unlocked_retryable(
+    payment_service: &BatcherPaymentService,
+    payment_service_fallback: &BatcherPaymentService,
+    addr: &Address,
+) -> Result<bool, RetryError<()>> {
+    if let Ok(unlock_block) = payment_service.user_unlock_block(*addr).call().await {
+        return Ok(unlock_block != U256::zero());
+    }
+    if let Ok(unlock_block) = payment_service_fallback
+        .user_unlock_block(*addr)
+        .call()
+        .await
+    {
+        return Ok(unlock_block != U256::zero());
+    }
+    warn!("Failed to get user locking state {:?}", addr);
+    Err(RetryError::Transient(()))
+}
+
+pub async fn get_gas_price_retryable(
+    eth_ws_provider: &Provider<Http>,
+    eth_ws_provider_fallback: &Provider<Http>,
+) -> Result<U256, RetryError<String>> {
+    if let Ok(gas_price) = eth_ws_provider
+        .get_gas_price()
+        .await
+        .inspect_err(|e| warn!("Failed to get gas price. Trying with fallback: {e:?}"))
+    {
+        return Ok(gas_price);
+    }
+
+    eth_ws_provider_fallback.get_gas_price().await.map_err(|e| {
+        warn!("Failed to get fallback gas price: {e:?}");
+        RetryError::Transient(e.to_string())
+    })
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::{
+        config::ECDSAConfig,
+        eth::{self, get_provider, payment_service::BatcherPaymentService},
+    };
+    use ethers::{
+        contract::abigen,
+        types::{Address, U256},
+        utils::{Anvil, AnvilInstance},
+    };
+    use std::str::FromStr;
+
+    abigen!(
+        BatcherPaymentServiceContract,
+        "../aligned-sdk/abi/BatcherPaymentService.json"
+    );
+
+    async fn setup_anvil(port: u16) -> (AnvilInstance, BatcherPaymentService) {
+        let anvil = Anvil::new()
+            .port(port)
+            .arg("--load-state")
+            .arg("../../contracts/scripts/anvil/state/alignedlayer-deployed-anvil-state.json")
+            .spawn();
+
+        let eth_rpc_provider = eth::get_provider(format!("http://localhost:{}", port))
+            .expect("Failed to get provider");
+
+        let payment_service_addr = String::from("0x7969c5eD335650692Bc04293B07F5BF2e7A673C0");
```

Comment:
_cries in hardcoded_

Reply:
Is reading the deployment output file an overkill? How difficult would it be?
They won't change frequently, but the test will fail when they do.

Reply:
How much parsing do they require? The file themselves may be `include_str!`'d to avoid dealing with files.

Reply:
Modified using ContractDeploymentOutput::new(), which is how we handle this in the batcher.